### PR TITLE
fix(npm): prefer importing esm from esm

### LINKED ISSUE #15593: Failed to `import XR from "npm:xregexp@5.1.0";`
```
error: Uncaught NotFound: No such file or directory (os error 2)
    at Object.Module._extensions..js (deno:ext/node/02_require.js:729:25)
    at Module.load (deno:ext/node/02_require.js:630:34)
    at Function.Module._load (deno:ext/node/02_require.js:487:14)
    at Module.require (deno:ext/node/02_require.js:652:21)
    at require (deno:ext/node/02_require.js:777:18)
    at Object.<anonymous> (file:///home/wojpawlik/.cache/deno/npm/registry.npmjs.org/@babel/runtime-corejs3/7.18.9/helpers/arrayWithHoles.js:1:278)
    at Object.<anonymous> (file:///home/wojpawlik/.cache/deno/npm/registry.npmjs.org/@babel/runtime-corejs3/7.18.9/helpers/arrayWithHoles.js:8:4)
    at Module._compile (deno:ext/node/02_require.js:713:36)
    at Object.Module._extensions..js (deno:ext/node/02_require.js:740:12)
    at Module.load (deno:ext/node/02_require.js:630:34)
```

### LINKED ISSUE #15642: npm URL throws: `Cannot read properties of undefined (reading 'filename')`
Happens with `g-i-s`:

```js
import gis from 'npm:g-i-s';
```

```
deno run -A --unstable .\main.js

error: Uncaught TypeError: Cannot read properties of undefined (reading 'filename')
    at Object.Module._extensions..js (deno:ext/node/02_require.js:734:64)
    at Module.load (deno:ext/node/02_require.js:630:34)
    at Function.Module._load (deno:ext/node/02_require.js:487:14)
    at Module.require (deno:ext/node/02_require.js:652:21)
    at require (deno:ext/node/02_require.js:776:18)
    at Object.<anonymous> (file:///C:/Users/GJZwiers/AppData/Local/deno/npm/registry.npmjs.org/cheerio/1.0.0-rc.12/lib/parsers/parse5-adapter.js:14:16)
    at Object.<anonymous> (file:///C:/Users/GJZwiers/AppData/Local/deno/npm/registry.npmjs.org/cheerio/1.0.0-rc.12/lib/parsers/parse5-adapter.js:68:4)
    at Module._compile (deno:ext/node/02_require.js:712:36)
    at Object.Module._extensions..js (deno:ext/node/02_require.js:739:12)
    at Module.load (deno:ext/node/02_require.js:630:34)
```

### PR DESCRIPTION
**Pull request by dsherret**
I just need to add some tests.

Also, this prefers using PathBuf over ModuleSpecifier in more places.

Closes #15593
Closes #15642

### CODE REVIEW

**Comment by bartlomieju** [comment]
This PR fixes https://github.com/denoland/deno/issues/15642

**Comment by bartlomieju** [comment]
This PR should fix https://github.com/denoland/deno/issues/15593 too, currently `op_require_try_self` returns a file URL (`file:///Users/ib/Library/Caches/deno/npm/registry.npmjs.org/@babel/runtime-corejs3/7.18.9/core-js/array/is-array.js`), when it should return a file path (`/Users/ib/Library/Caches/deno/npm/registry.npmjs.org/@babel/runtime-corejs3/7.18.9/core-js/array/is-array.js`)

File: cli/node/mod.rs [hunk 1]
```diff
@@ -412,24 +416,21 @@ pub fn node_resolve_npm_reference(
...
-  let maybe_url = package_config_resolve(
+  let resolved_path = package_config_resolve(
     &reference
       .sub_path
       .as_ref()
       .map(|s| format!("./{}", s))
       .unwrap_or_else(|| ".".to_string()),
     &package_folder,
     npm_resolver,
+    NodeModuleKind::Esm,
```
Comment (bartlomieju) [inline cli/node/mod.rs:427]: Why is it always ESM?

**Review by bartlomieju — APPROVED** [review]
LGTM

File: cli/node/mod.rs (hunk 1 above)
Comment (dsherret) [reply cli/node/mod.rs:427]: It’s the referrer kind, which is an npm package reference, which is deno code, so esm.

**Pull request merged**